build/lib/src/generate/run_builder.dart

```diff
@@ -26,12 +26,16 @@ import 'expected_outputs.dart';
 /// automatically disposed of (its up to the caller to dispose of it later). If
 /// one is not provided then one will be created and disposed at the end of
 /// this function call.
+///
+/// If [removedDependencies] is provided an entry will be added for each input
+/// which gives all the removed dependencies for that input.
 Future<void> runBuilder(Builder builder, Iterable<AssetId> inputs,
     AssetReader reader, AssetWriter writer, Resolvers resolvers,
     {Logger logger,
     ResourceManager resourceManager,
     String rootPackage,
-    StageTracker stageTracker = NoOpStageTracker.instance}) async {
+    StageTracker stageTracker = NoOpStageTracker.instance,
+    Map<AssetId, Iterable<AssetId>> removedDependencies}) async {
```

Comment:
This is also a weird api that is a side effect of putting this on the `BuildStep` class - this method can create multiple `BuildStep` instances and doesn't return them. Because of that I end up allowing you to pass in this map to get the information back out :(.

Reply:
I'm not a fan of "output collections" like this. Can we pass a callback? If we pass it in to the constructor of `BuildStepImpl` it would also obviate the need for adding an API on `buildStep` that is only useful to the build system.

```
...
void Function(AssetId, Set<AssetId>) reportUnusedAssets

...

BuildStepImpl(input, outputs,...
  reportUnusedAssets: (ids) => reportUnusedAssets(input, ids),
);
```

Reply:
Ya I like that more - and it also removes the need for the casts to `Set`.

Actually I think I will remove the internal set entirely and move that out to the client code as well, by making this `void Function reportUnusedAsset(AssetId, AssetId)`?

Reply:
this is done now